pkg/router/virtualservice.go

```diff
@@ -29,15 +47,83 @@ func (v *VirtualService) Clear(s Shift) error {
 		var cleanedRules []*v1alpha3.HTTPRoute
 		cleanedRules = []*v1alpha3.HTTPRoute{}
 
-		logger.Info(fmt.Sprintf("removing all virtualservice '%s' rules except the master-route one (Regex: .+)", vs.Name), v.TrackingId)
-		for httpKey, httpValue := range vs.Spec.Http {
-			for _, matchValue := range httpValue.Match {
-				if matchValue.Uri.GetRegex() == ".+" {
-					cleanedRules = append(cleanedRules, vs.Spec.Http[httpKey])
+		if m == "hard" {
+			logger.Info(fmt.Sprintf("triggering hard clear. Removing all virtualService '%s' rules except the master-route one (Regex: .+)", vs.Name), v.TrackingId)
+			for httpKey, httpValue := range vs.Spec.Http {
+				for _, matchValue := range httpValue.Match {
+					anyPrefix, _ := regexp.MatchString(`.+`, matchValue.Uri.GetPrefix())
+					if anyPrefix {
+						cleanedRules = append(cleanedRules, vs.Spec.Http[httpKey])
+					}
+					if matchValue.Uri.GetRegex() == ".+" {
+						cleanedRules = append(cleanedRules, vs.Spec.Http[httpKey])
+					}
 				}
 			}
 		}
 
+		if m == "soft" {
+			logger.Info(fmt.Sprintf("triggering soft clear for virtualService '%s'", vs.Name), v.TrackingId)
+			for httpKey, httpValue := range vs.Spec.Http {
+				// append canary rules without pods associated - based on destinationRules
+				for _, routeValue := range httpValue.Route {
+					// subset can be empty and won't be removed
+					if routeValue.Destination.Subset == "" {
+						logger.Debug("including a 'joker' route which does not have any subset configured", v.TrackingId)
+						cleanedRules = append(cleanedRules, vs.Spec.Http[httpKey])
+					}
+
+					if routeValue.Destination.Subset != "" {
+						for _, d := range dss.DList.Items {
+							for _, subset := range d.Spec.Subsets {
```

Comment:
This for structure could be separated from the immediately upper for structure by ranging an array mounted by it considering the "routeValue.Destination.Subset != """ condition.